pyanaconda/modules/boss/boss.py

```diff
@@ -0,0 +1,161 @@
+        services = ["%s.%s" % (constants.DBUS_MODULE_NAMESPACE, m)
+                    for m in self.modules]
+        services.extend(["%s.%s" % (constants.DBUS_ADDON_NAMESPACE, a)
+                         for a in self.addons])
+        return services
+
+    @property
+    def running_module_services(self):
+        # For our purpose, just this
+        return self._started_module_services
+
+    def EchoString(self, s: Str) -> Str:
+        """returns whatever is passed to it"""
+        log.debug(s)
+        return s
+
+    def Quit(self):
+        """removes the object from the DBUS connection and exits"""
+        log.debug("sending Quit to all modules and addons")
+        for service in self.running_module_services:
+            module = self._bus.get(service)
+            # TODO: async
+            try:
+                module.Quit()
+            except Exception:
+                log.exception("Boss shutdown failed")
+        self._loop.quit()
+
+    def modules_starting_finished(self):
+        # TODO: do we need some synchronization here?
+        return set(self._started_module_services + self._failed_module_services) == set(self.required_module_services)
+
+    def _start_service_cb(self, service, returned=None, error=None):
+        """callback for dbus.StartServiceByName"""
+        if error:
+            log.debug("%s error: %s" % (service,error))
+            self._failed_module_services.append(service)
+        elif returned:
+            if returned == DBUS_START_REPLY_SUCCESS:
+                log.debug("%s started successfully, returned: %s)" % (service, returned))
+                self._started_module_services.append(service)
+            else:
+                log.debug("%s failed to start, returned: %s" % (service, returned))
+                self._failed_module_services.append(service)
+
+        self.check_modules_started()
+
+    def check_modules_started(self):
+        if self.modules_starting_finished():
+            log.debug("modules starting finished, running: %s failed: %s" %
+                      (self._started_module_services,
+                       self._failed_module_services))
+            for service in self._started_module_services:
+                log.debug("SERVICE: %s", service)
+                module = self._bus.get(service)
+                module.EchoString("Boss told me that all required modules were started: %s or failed: %s." %
+                                  (self._started_module_services, self._failed_module_services))
+            return True
+        else:
+            return False
+
+    def _start_modules(self):
+        """starts anaconda modules (including addons)"""
+        log.debug("started")
+
+        # Check that no modules are running
+        for service in self.required_module_services:
+            if self._dbus.NameHasOwner(service):
+                log.error("service %s has unexpected owner" % service)
+
+        for service in self.required_module_services:
```

Comment:
I think this is deliberate (this is also taken from Radeks example mostly verbatim) so that you first report all errors (if any) and only then try actually starting the modules.